File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatadetail/WeatherDataDetailFragment.java
package pl.mosenko.sunnypodlaskie.mvp.weatherdatadetail;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.hannesdorfmann.mosby3.mvp.MvpFragment;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import pl.mosenko.sunnypodlaskie.BuildConfig;
import pl.mosenko.sunnypodlaskie.R;

/**
 * Created by syk on 20.05.17.
 */

public class WeatherDataDetailFragment extends MvpFragment<WeatherDataDetailContract.View, WeatherDataDetailContract.Presenter> implements WeatherDataDetailContract.View {

    public static final String TAG = WeatherDataDetailFragment.class.getSimpleName();
    public static final String ARG_WEATHER_DATA_ID = "weather_data_id";

    @BindView(R.id.WeatherDataDetailFragment_city_detail)
    TextView mCityDetailTV;
    @BindView(R.id.WeatherDataDetailFragment_weather_icon_detail)
    ImageView mIconDetailIV;
    @BindView(R.id.WeatherDataDetailFragment_temperature_detail)
    TextView mTemperatureTV;
    @BindView(R.id.WeatherDataDetailFragment_weather_description_detail)
    TextView mWeatherDescriptionTV;
    @BindView(R.id.WeatherDataDetailFragment_pressure_detail)
    TextView mPressure;
    @BindView(R.id.WeatherDataDetailFragment_wind_detail)
    TextView mWindDetails;
    @BindView(R.id.WeatherDataDetailFragment_humidity_detail)
    TextView mHumidityDetail;
    @BindView(R.id.WeatherDataDetailFragment_sunrise_detail)
    TextView mSunrise;
    @BindView(R.id.WeatherDataDetailFragment_sunset_detail)
    TextView mSunset;

    private Unbinder unbinder;

    public WeatherDataDetailFragment() {
    }

    @Override
    public WeatherDataDetailContract.Presenter createPresenter() {
        return new WeatherDataDetailPresenterImpl();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflateFragment(inflater, container);
        bindGraphicalComponents(rootView);
        return rootView;
    }

    private View inflateFragment(LayoutInflater inflater, ViewGroup container) {
        return inflater.inflate(R.layout.fragment_weather_data_details, container, false);
    }

    private void bindGraphicalComponents(View rootView) {
        unbinder = ButterKnife.bind(this, rootView);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getPresenter().onViewCreated(getArguments());
    }

    @Override
    public void loadData(WeatherDataDetailPresentationModel weatherDataDetailPresentationModel) {
        mCityDetailTV.setText(weatherDataDetailPresentationModel.getTitleDetails());
        mTemperatureTV.setText(weatherDataDetailPresentationModel.getTemperature());
        mWeatherDescriptionTV.setText(weatherDataDetailPresentationModel.getDescription());
        mIconDetailIV.setImageResource(weatherDataDetailPresentationModel.getIconResource());
        mPressure.setText(weatherDataDetailPresentationModel.getPressure());
        mWindDetails.setText(weatherDataDetailPresentationModel.getWindDetails());
        mHumidityDetail.setText(weatherDataDetailPresentationModel.getHumidity());
        mSunrise.setText(weatherDataDetailPresentationModel.getSunrise());
        mSunset.setText(weatherDataDetailPresentationModel.getSunset());
    }

    @Override
    public void showError(Throwable throwable) {
        if (BuildConfig.DEBUG) {
            Log.e(TAG, throwable.getMessage(), throwable);
        }
        Toast.makeText(getContext(), R.string.weather_details_error_message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatadetail/WeatherDataDetailFragment.kt
package pl.mosenko.sunnypodlaskie.mvp.weatherdatadetail

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.navArgs
import com.google.android.material.snackbar.Snackbar
import org.koin.androidx.viewmodel.ext.android.viewModel
import pl.mosenko.sunnypodlaskie.BuildConfig
import pl.mosenko.sunnypodlaskie.R
import pl.mosenko.sunnypodlaskie.databinding.FragmentWeatherDataDetailsBinding
import pl.mosenko.sunnypodlaskie.mvp.EventObserver

/**
 * Created by syk on 20.05.17.
 */
class WeatherDataDetailFragment : Fragment() {

    private val args: WeatherDataDetailFragmentArgs by navArgs()
    private var _binding: FragmentWeatherDataDetailsBinding? = null
    private val binding get() = _binding!!
    val viewModel: WeatherDataDetailViewModel by viewModel()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentWeatherDataDetailsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (arguments == null || requireArguments().isEmpty) {
            viewModel.loadWeatherDataDetails(null)
        } else {
            viewModel.loadWeatherDataDetails(args.city)
        }
        setupObservers()
    }

    private fun setupObservers() {
        viewModel.weatherDataDetails.observe(viewLifecycleOwner, {
            if (it != null) {
                showData(it)
            }
        })
        viewModel.errorEvent.observe(viewLifecycleOwner, EventObserver {
            if (BuildConfig.DEBUG) {
                Log.e(TAG, it.message, it)
            }
            Snackbar.make(
                binding.root,
                R.string.weather_details_error_message,
                Snackbar.LENGTH_LONG
            ).show()
        })
    }

    private fun showData(weatherDataDetailPresentationModel: WeatherDataDetailPresentationModel) {
        binding.incPrimaryInfo.apply {
            tvCityDetail.text = weatherDataDetailPresentationModel.titleDetails
            tvTemperatureDetail.text = weatherDataDetailPresentationModel.temperature
            tvWeatherDescriptionDetail.text = weatherDataDetailPresentationModel.description
            ivWeatherIconDetail.setImageResource(weatherDataDetailPresentationModel.iconResource)
        }
        binding.incExtraInfo.apply {
            tvPressureDetail.text = weatherDataDetailPresentationModel.pressure
            tvWindDetail.text = weatherDataDetailPresentationModel.windDetails
            tvHumidityDetail.text = weatherDataDetailPresentationModel.humidity
            tvSunriseDetail.text = weatherDataDetailPresentationModel.sunrise
            tvSunsetDetail.text = weatherDataDetailPresentationModel.sunset
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        val TAG: String = WeatherDataDetailFragment::class.java.simpleName
        const val ARG_WEATHER_DATA_CITY: String = "weatherDataId"
    }
}
